File: missing-default-combination-zero-price/node/fix-default-combination.js
/**
 * Find and repair PrestaShop products whose default combination pointer is stale.
 *
 * A product with combinations shows its headline price by resolving id_default_combination
 * to one specific combination row. When that pointer is 0, blank, or names a combination
 * that was deleted, deactivated, or belongs to a different product, the price lookup has
 * nothing valid to read and the product displays a price of zero even though its other
 * combinations have real prices.
 *
 * This script lists products, pulls each one's live combinations, and checks whether the
 * stored id_default_combination resolves to an active combination that still belongs to
 * that product. When it does not and an eligible combination exists, it repairs the
 * pointer to the cheapest eligible one. When no eligible combination exists at all, it
 * flags the product for a human instead of guessing. The only write is a PUT on the
 * product's own id_default_combination field; combination rows are never modified.
 *
 * Guide: https://www.allanninal.dev/prestashop/missing-default-combination-zero-price/
 */
import { pathToFileURL } from "node:url";

const PRESTASHOP_URL = (process.env.PRESTASHOP_URL || "https://demo.example.com").replace(/\/+$/, "");
const PRESTASHOP_WS_KEY = process.env.PRESTASHOP_WS_KEY || "WSKEYDUMMY";
const DRY_RUN = (process.env.DRY_RUN || "true").toLowerCase() === "true";

function basicAuthHeader() {
  return "Basic " + Buffer.from(`${PRESTASHOP_WS_KEY}:`).toString("base64");
}

/**
 * Pure decision function, no I/O.
 *
 * idProduct: the product being checked.
 * currentDefaultId: the product's stored id_default_combination value.
 * combinations: the product's live combinations, each an object with id, id_product,
 *   active, and price.
 *
 * Returns a decision object describing what to do. All HTTP calls happen in the caller.
 */
export function decideDefaultCombination(idProduct, currentDefaultId, combinations) {
  const eligible = (c) => String(c.id_product) === String(idProduct) && String(c.active ?? "0") === "1";

  const liveIds = new Set(combinations.filter(eligible).map((c) => String(c.id)));
  const isValid = !["", "0", "null", "undefined"].includes(String(currentDefaultId ?? "")) &&
    liveIds.has(String(currentDefaultId));

  if (isValid) {
    return {
      action: "none",
      reason: "default combination is active and belongs to the product",
      targetId: null,
    };
  }

  const eligibleCombos = combinations.filter(eligible);
  if (eligibleCombos.length === 0) {
    return {
      action: "flag",
      reason: "no active combination belongs to this product",
      targetId: null,
    };
  }

  const cheapest = eligibleCombos.reduce((best, c) =>
    Number(c.price || 0) < Number(best.price || 0) ? c : best
  );
  return {
    action: "repair",
    reason: "default combination missing or invalid, replacing with cheapest active one",
    targetId: cheapest.id,
  };
}

async function apiGet(path, params = {}) {
  const url = new URL(`${PRESTASHOP_URL}/api/${path}`);
  url.searchParams.set("output_format", "JSON");
  for (const [key, value] of Object.entries(params)) url.searchParams.set(key, value);
  const res = await fetch(url, { headers: { Authorization: basicAuthHeader() } });
  if (!res.ok) throw new Error(`PrestaShop ${res.status} on GET ${path}`);
  return res.json();
}

async function apiPut(path, resourceKey, body) {
  const url = new URL(`${PRESTASHOP_URL}/api/${path}`);
  url.searchParams.set("output_format", "JSON");
  const res = await fetch(url, {
    method: "PUT",
    headers: { Authorization: basicAuthHeader(), "Content-Type": "application/json" },
    body: JSON.stringify({ [resourceKey]: body }),
  });
  if (!res.ok) throw new Error(`PrestaShop ${res.status} on PUT ${path}`);
  return res.json();
}

async function listProductsWithCombinations(limit = 100) {
  const data = await apiGet("products", { display: "full", limit });
  const products = data.products || [];
  return products.filter((p) => String(p.id_default_combination ?? "0") !== "");
}

async function listCombinations(idProduct) {
  const data = await apiGet("combinations", {
    "filter[id_product]": idProduct,
    display: "full",
  });
  return data.combinations || [];
}

async function repairDefaultCombination(product, targetCombinationId) {
  const body = { ...product, id_default_combination: targetCombinationId };
  return apiPut(`products/${product.id}`, "product", body);
}

export async function run() {
  let flagged = 0;
  let repaired = 0;
  for (const product of await listProductsWithCombinations()) {
    const idProduct = product.id;
    const currentDefaultId = product.id_default_combination;

    const combinations = await listCombinations(idProduct);
    const decision = decideDefaultCombination(idProduct, currentDefaultId, combinations);

    if (decision.action === "none") continue;

    flagged++;
    console.warn(
      `Product ${idProduct} current id_default_combination=${currentDefaultId} ` +
        `action=${decision.action} reason=${decision.reason} target_id=${decision.targetId}`
    );

    if (decision.action === "repair" && !DRY_RUN) {
      await repairDefaultCombination(product, decision.targetId);
      repaired++;
      console.log(`Repaired product ${idProduct} id_default_combination=${decision.targetId}.`);
    }
  }
  console.log(`Done. ${flagged} product(s) flagged, ${repaired} repaired.`);
}

if (import.meta.url === pathToFileURL(process.argv[1]).href) {
  run().catch((err) => { console.error(err); process.exit(1); });
}

File: missing-default-combination-zero-price/python/fix_default_combination.py
"""Find and repair PrestaShop products whose default combination pointer is stale.

A product with combinations shows its headline price by resolving id_default_combination
to one specific combination row. When that pointer is 0, blank, or names a combination
that was deleted, deactivated, or belongs to a different product, the price lookup has
nothing valid to read and the product displays a price of zero even though its other
combinations have real prices.

This script lists products, pulls each one's live combinations, and checks whether the
stored id_default_combination resolves to an active combination that still belongs to
that product. When it does not and an eligible combination exists, it repairs the
pointer to the cheapest eligible one. When no eligible combination exists at all, it
flags the product for a human instead of guessing. The only write is a PUT on the
product's own id_default_combination field; combination rows are never modified.

Run after bulk combination edits, imports, or product duplication. Safe to run again
and again.

Guide: https://www.allanninal.dev/prestashop/missing-default-combination-zero-price/
"""
import os
import logging
import requests

logging.basicConfig(level=logging.INFO, format="%(asctime)s %(levelname)s %(message)s")
log = logging.getLogger("fix_default_combination")

PRESTASHOP_URL = os.environ["PRESTASHOP_URL"].rstrip("/")
PRESTASHOP_WS_KEY = os.environ["PRESTASHOP_WS_KEY"]
DRY_RUN = os.environ.get("DRY_RUN", "true").lower() == "true"
AUTH = (PRESTASHOP_WS_KEY, "")


def decide_default_combination(id_product, current_default_id, combinations):
    """Pure decision function, no I/O.

    id_product: the product being checked.
    current_default_id: the product's stored id_default_combination value.
    combinations: the product's live combinations, each a dict with id, id_product,
        active, and price.

    Returns a decision dict describing what to do. All HTTP calls happen in the caller.
    """
    def eligible(c):
        return (
            str(c.get("id_product")) == str(id_product)
            and str(c.get("active", "0")) == "1"
        )

    live_ids = {str(c["id"]) for c in combinations if eligible(c)}
    is_valid = str(current_default_id) not in ("", "0", "None") and str(current_default_id) in live_ids

    if is_valid:
        return {
            "action": "none",
            "reason": "default combination is active and belongs to the product",
            "target_id": None,
        }

    eligible_combos = [c for c in combinations if eligible(c)]
    if not eligible_combos:
        return {
            "action": "flag",
            "reason": "no active combination belongs to this product",
            "target_id": None,
        }

    cheapest = min(eligible_combos, key=lambda c: float(c.get("price", 0) or 0))
    return {
        "action": "repair",
        "reason": "default combination missing or invalid, replacing with cheapest active one",
        "target_id": cheapest["id"],
    }


def api_get(path, params=None):
    params = dict(params or {})
    params["output_format"] = "JSON"
    r = requests.get(f"{PRESTASHOP_URL}/api/{path}", params=params, auth=AUTH, timeout=30)
    r.raise_for_status()
    return r.json()


def api_put(path, resource_key, body):
    r = requests.put(
        f"{PRESTASHOP_URL}/api/{path}",
        params={"output_format": "JSON"},
        auth=AUTH,
        json={resource_key: body},
        timeout=30,
    )
    r.raise_for_status()
    return r.json()


def list_products_with_combinations(limit=100):
    data = api_get("products", params={"display": "full", "limit": limit})
    products = data.get("products") or []
    return [p for p in products if str(p.get("id_default_combination", "0")) != ""]


def list_combinations(id_product):
    data = api_get("combinations", params={
        "filter[id_product]": id_product,
        "display": "full",
    })
    return data.get("combinations") or []


def repair_default_combination(product, target_combination_id):
    body = dict(product)
    body["id_default_combination"] = target_combination_id
    return api_put(f"products/{product['id']}", "product", body)


def run():
    flagged = 0
    repaired = 0
    for product in list_products_with_combinations():
        id_product = product.get("id")
        current_default_id = product.get("id_default_combination")

        combinations = list_combinations(id_product)
        decision = decide_default_combination(id_product, current_default_id, combinations)

        if decision["action"] == "none":
            continue

        flagged += 1
        log.warning(
            "Product %s current id_default_combination=%s action=%s reason=%s target_id=%s",
            id_product, current_default_id, decision["action"], decision["reason"], decision["target_id"],
        )

        if decision["action"] == "repair" and not DRY_RUN:
            repair_default_combination(product, decision["target_id"])
            repaired += 1
            log.info("Repaired product %s id_default_combination=%s.", id_product, decision["target_id"])

    log.info("Done. %d product(s) flagged, %d repaired.", flagged, repaired)


if __name__ == "__main__":
    run()
